File: modules/cms/models/purchase.go
package models

import (
	"aldev/modules/global/models"
	"time"

	"github.com/google/uuid"
)

type Purchase struct {
	models.BaseModel
	PurchaseNumber   *string           `json:"purchase_number" gorm:"type:varchar(100);unique;index;not null"`
	PurchaseDate     *time.Time        `json:"purchase_date" gorm:"not null"`
	Status           *string           `json:"status" gorm:"type:varchar(50);default:'draft'"`
	IsClear          *bool             `json:"is_clear" gorm:"default:false;not null"`
	PrincipleID      *uuid.UUID        `json:"principle_id" gorm:"type:uuid;not null"`
	Principle        *Principle        `json:"principle,omitempty" gorm:"foreignKey:PrincipleID"`
	TotalPrice       *float64          `json:"total_price" gorm:"type:decimal(15,2);not null"`
	Notes            *string           `json:"notes,omitempty" gorm:"type:text"`
	Attachments      *string           `json:"attachments,omitempty" gorm:"type:text"`
	PurchaseProducts []PurchaseProduct `json:"purchase_products,omitempty" gorm:"foreignKey:PurchaseID"`
}

type PurchaseProduct struct {
	models.BaseModel
	PurchaseID *uuid.UUID `json:"purchase_id" gorm:"type:uuid;not null"`
	ProductID  *uuid.UUID `json:"product_id" gorm:"type:uuid;not null"`
	Product    *Product   `json:"product,omitempty" gorm:"foreignKey:ProductID"`
	Qty        *int       `json:"qty" gorm:"not null"`

	// For individual tracking products
	LengthPerItem   *float64 `json:"length_per_item,omitempty" gorm:"type:decimal(15,3)"`
	Width           *float64 `json:"width,omitempty" gorm:"type:decimal(10,2)"`
	MeasurementUnit *string  `json:"measurement_unit,omitempty" gorm:"type:varchar(20)"`

	PurchasePrice *float64 `json:"purchase_price" gorm:"type:decimal(15,2);not null"`
	Subtotal      *float64 `json:"subtotal" gorm:"type:decimal(15,2);not null"`
}

File: src/types/purchase.ts
import * as z from 'zod';
import type { Product } from './product';
import type { Principle } from './principle';

// Purchase Item Schema (for form)
export const PurchaseItemSchema = z.object({
   product_id: z.string().min(1, { message: 'Product is required!' }),
   qty: z.coerce.number().min(1, { message: 'Quantity must be at least 1!' }),
   purchase_price: z.coerce.number().min(0, { message: 'Purchase Price must be at least 0!' }),

   // For individual tracking products
   length_per_item: z.coerce.number().optional(),
   width: z.coerce.number().optional(),
   measurement_unit: z.string().optional(),
});

export type PurchaseItemFormType = z.infer<typeof PurchaseItemSchema>;

// Purchase Schema (for form)
export const PurchaseSchema = z.object({
   principle_id: z.string().min(1, { message: 'Supplier is required!' }),
   purchase_date: z.string().min(1, { message: 'Purchase Date is required!' }),
   notes: z.string().optional(),
   items: z.array(PurchaseItemSchema).min(1, { message: 'At least one product is required!' }),
});

export type PurchaseFormType = z.infer<typeof PurchaseSchema>;

// Backend interfaces
export interface PurchaseProduct {
   id?: string;
   created_at?: string;
   updated_at?: string;
   purchase_id?: string;
   product_id?: string;
   product?: Product;
   qty?: number;
   purchase_price?: number;
   subtotal?: number;

   // For individual tracking products
   length_per_item?: number;
   width?: number;
   measurement_unit?: string;
}

export interface Purchase {
   id?: string;
   created_at?: string;
   updated_at?: string;
   purchase_number?: string;
   purchase_date?: string;
   status?: 'draft' | 'completed' | 'cancelled';
   is_clear?: boolean;
   principle_id?: string;
   principle?: Principle;
   total_price?: number;
   notes?: string;
   attachments?: string;
   purchase_products?: PurchaseProduct[];
   // Alias for form compatibility
   items?: PurchaseProduct[];
}
